src/libraries/System.Private.CoreLib/src/System/IO/FileStream.cs

```diff
@@ -223,38 +172,23 @@ public FileStream(string path, FileMode mode, FileAccess access, FileShare share
             if ((access & FileAccess.Read) != 0 && mode == FileMode.Append)
                 throw new ArgumentException(SR.Argument_InvalidAppendMode, nameof(access));
 
-            string fullPath = Path.GetFullPath(path);
-
-            _path = fullPath;
-            _access = access;
-            _bufferLength = bufferSize;
-
-            if ((options & FileOptions.Asynchronous) != 0)
-                _useAsyncIO = true;
-
             if ((access & FileAccess.Write) == FileAccess.Write)
             {
                 SerializationInfo.ThrowIfDeserializationInProgress("AllowFileWrites", ref s_cachedSerializationSwitch);
             }
 
-            _fileHandle = OpenHandle(mode, share, options);
-
-            try
+            if ((options & FileOptions.Asynchronous) != 0)
             {
-                Init(mode, share, path);
+                _impl = new FileStreamImpl(path, mode, access, share, bufferSize, options);
             }
-            catch
+            else
             {
-                // If anything goes wrong while setting up the stream, make sure we deterministically dispose
-                // of the opened handle.
-                _fileHandle.Dispose();
-                _fileHandle = null!;
-                throw;
+                _impl = new FileStreamImpl(path, mode, access, share, bufferSize, options);
             }
         }
 
         [Obsolete("This property has been deprecated.  Please use FileStream's SafeFileHandle property instead.  https://go.microsoft.com/fwlink/?linkid=14202")]
-        public virtual IntPtr Handle => SafeFileHandle.DangerousGetHandle();
+        public virtual IntPtr Handle => _impl.Handle;
```

Comment:
I'd have expected some things you could avoid duplicating between the implementations, e.g. I'd expect they'd all need the SafeFileHandle and in the same way, such that it could live as a field on the FileStream, and every impl would just have a field pointing to its parent FileStream.